### PR TITLE
Replace Hash with Rack::Header/Rack::Utils::HeaderHash for Grape Headers

### PR DESCRIPTION
**Pull request by dhruvCW**
This closes https://github.com/ruby-grape/grape/issues/2396

### CODE REVIEW

**Comment by ericproulx** [comment]
I think there's this one left [here](https://github.com/ruby-grape/grape/blob/5121279107b3b905203f48a94a8a2069659ec519/lib/grape/endpoint.rb#L241)

**Comment by dhruvCW** [comment]
> I think there's this one left [here](https://github.com/ruby-grape/grape/blob/5121279107b3b905203f48a94a8a2069659ec519/lib/grape/endpoint.rb#L241). Thanks for doing this :)

ah thanks 😅 missed that, should be fixed now 👍

File: CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +18,7 @@
 * [#2408](https://github.com/ruby-grape/grape/pull/2408): Fix params method redefined warnings - [@ericproulx](https://github.com/ericproulx).
 * [#2410](https://github.com/ruby-grape/grape/pull/2410): Gem deprecations will raise a DeprecationWarning in specs - [@ericproulx](https://github.com/ericproulx).
 * [#2389](https://github.com/ruby-grape/grape/pull/2389): Remove rack-accept dependency - [@ericproulx](https://github.com/ericproulx).
+* [#2425](https://github.com/ruby-grape/grape/pull/2425): Replace `{}` with `Rack::Header` (for rack 3.0 and greater) or Rack::Utils::HeaderHash` (for rack less than 3.0) in Grape::DSL::Headers - [@dhruvCW](https://github.com/dhruvCW).
```
Comment (dblock) [inline CHANGELOG.md:21]: Quote these consistently, `Grape::DSL::Headers`, maybe shorten.

Replace `{}` with `Rack::Header` or `Rack::Utils::HeaderHash`.

or if you really want versions here

Replace `{}` with `Rack::Header` in Rack 3.0+ and Rack::Utils::HeaderHash` for Rack 1.x and 2.x.

File: lib/grape/dsl/headers.rb [hunk 2]
```diff
@@ -1,5 +1,6 @@
 # frozen_string_literal: true
 
+require 'grape/util/header'
```
Comment (dblock) [inline lib/grape/dsl/headers.rb:3]: Add a blank line after.

File: lib/grape/endpoint.rb [hunk 3]
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require 'grape/util/header'
```
Comment (dblock) [inline lib/grape/endpoint.rb:3]: Can we consolidate this require elsewhere once?

File: lib/grape/util/header.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module Grape
+  module Util
+    begin
+      # For `Rack::Headers` (Rack 3+):
+      require 'rack/headers'
+      Header = Rack::Headers
+    rescue LoadError
```
Comment (dblock) [inline lib/grape/util/header.rb:9]: Don't rely on an exception at runtime, use `Gem::Version.new(Rack.release) >= Gem::Version.new('3')`.

**Review by dblock — CHANGES_REQUESTED** [review]
Thanks! Some asks below.

File: CHANGELOG.md (hunk 1 above)
Comment (dhruvCW) [reply CHANGELOG.md]: yup the one without version sounds better 😅 replaced.

File: lib/grape/util/header.rb (hunk 4 above)
Comment (dhruvCW) [reply lib/grape/util/header.rb]: awesome, replaced 👍

File: lib/grape/dsl/headers.rb (hunk 2 above)
Comment (dhruvCW) [reply lib/grape/dsl/headers.rb]: removed (not required anymore).

File: lib/grape/endpoint.rb (hunk 3 above)
Comment (dhruvCW) [reply lib/grape/endpoint.rb]: removed (not required anymore)

**Comment by dblock** [comment]
Great, merging. I don't think we need anything for UPGRADING.md, but if you disagree, please PR.

**Review by dblock — APPROVED** [review]

**Pull request merged**